File: java/BinarySearch/Minimized_Maximum_of_Products_Distributed_to_Any_Store_2064.java
package java.BinarySearch;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Minimized_Maximum_of_Products_Distributed_to_Any_Store_2064 {
    // same as koko eating banana same code
    public int minimizedMaximum(int n, int[] quantities) {
        int start = 1;
        int end = 0;
        for (int i : quantities)
            end = Math.max(end, i);
        int ans = end;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            int count = 0;
            for (int product : quantities) {
                if (product > mid)
                    count += (int) Math.ceil(product * 1.0 / mid);
                else
                    count += 1;
                if (count > n)
                    break;
            }
            if (count > n) {
                start = mid + 1;
            } else {
                ans = Math.min(ans, mid);
                end = mid - 1;
            }
        }
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Minimized_Maximum_of_Products_Distributed_to_Any_Store_2064 obj = new Minimized_Maximum_of_Products_Distributed_to_Any_Store_2064();
            int nums[] = { 11, 6 };

            op.println(obj.minimizedMaximum(6, nums));

        }
    }

}

File: Code/BinarySearch/Minimized_Maximum_of_Products_Distributed_to_Any_Store_2064.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    // same as koko eating banana
    int minimizedMaximum(int n, vector<int> quantities)
    {
        int start = 1;
        int end = *max_element(quantities.begin(), quantities.end());
        int ans = end;
        while (start <= end)
        {
            int mid = start + (end - start) / 2;
            int count = 0;
            for (int product : quantities)
            {
                if (product > mid)
                    count += ceil(product * 1.0 / mid);
                else
                    count += 1;
                if (count > n)
                    break;
            }
            if (count > n)
            {
                start = mid + 1;
            }
            else
            {
                ans = min(ans, mid);
                end = mid - 1;
            }
        }
        return ans;
    }
};

int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.minimizedMaximum(6, {11, 6});
    cout << ans;
    return 0;
}